Source/Core/InputCommon/ControllerInterface/XInput/XInput.cpp

```diff
@@ -10,43 +10,120 @@
+    {"Button X", XINPUT_GAMEPAD_X},
+    {"Button Y", XINPUT_GAMEPAD_Y},
+    {"Pad N", XINPUT_GAMEPAD_DPAD_UP},
+    {"Pad S", XINPUT_GAMEPAD_DPAD_DOWN},
+    {"Pad W", XINPUT_GAMEPAD_DPAD_LEFT},
+    {"Pad E", XINPUT_GAMEPAD_DPAD_RIGHT},
+    {"Start", XINPUT_GAMEPAD_START},
+    {"Back", XINPUT_GAMEPAD_BACK},
+    {"Shoulder L", XINPUT_GAMEPAD_LEFT_SHOULDER},
+    {"Shoulder R", XINPUT_GAMEPAD_RIGHT_SHOULDER},
+    {"Guide", XINPUT_GAMEPAD_GUIDE},
+    {"Thumb L", XINPUT_GAMEPAD_LEFT_THUMB},
+    {"Thumb R", XINPUT_GAMEPAD_RIGHT_THUMB},
+}};
+
+static constexpr std::array named_triggers{"Trigger L", "Trigger R"};
+
+static constexpr std::array named_axes{"Left X", "Left Y", "Right X", "Right Y"};
+
+static constexpr std::array named_motors{"Motor L", "Motor R"};
+
+class Button final : public Core::Device::Input
+{
+public:
+  Button(u8 index, const WORD& buttons) : m_buttons(buttons), m_index(index) {}
+  std::string GetName() const override { return named_buttons[m_index].name; }
+  ControlState GetState() const override
+  {
+    return (m_buttons & named_buttons[m_index].bitmask) > 0;
+  }
+
+private:
+  const WORD& m_buttons;
+  const u8 m_index;
+};
+
+class Axis final : public Core::Device::Input
+{
+public:
+  Axis(u8 index, const SHORT& axis, SHORT range) : m_axis(axis), m_range(range), m_index(index) {}
+  std::string GetName() const override
+  {
+    return std::string(named_axes[m_index]) + (m_range < 0 ? '-' : '+');
+  }
+  ControlState GetState() const override { return ControlState(m_axis) / m_range; }
+
+private:
+  const SHORT& m_axis;
+  const SHORT m_range;
+  const u8 m_index;
+};
+
+class Trigger final : public Core::Device::Input
+{
+public:
+  Trigger(u8 index, const BYTE& trigger, BYTE range)
+      : m_trigger(trigger), m_range(range), m_index(index)
+  {
+  }
+  std::string GetName() const override { return named_triggers[m_index]; }
+  ControlState GetState() const override { return ControlState(m_trigger) / m_range; }
+
+private:
+  const BYTE& m_trigger;
+  const BYTE m_range;
+  const u8 m_index;
+};
+
+class Motor final : public Core::Device::Output
+{
+public:
+  Motor(u8 index, Device* parent, WORD& motor, WORD range)
+      : m_motor(motor), m_range(range), m_index(index), m_parent(parent)
+  {
+  }
+
+  std::string GetName() const override { return named_motors[m_index]; }
+  void SetState(ControlState state) override
+  {
+    const auto old_value = m_motor;
+    m_motor = (WORD)(state * m_range);
+
+    // Only update if the state changed.
+    if (m_motor != old_value)
+      m_parent->UpdateMotors();
+  }
+
+private:
+  WORD& m_motor;
+  const WORD m_range;
+  const u8 m_index;
+  Device* m_parent;
+};
+
+class Battery final : public Core::Device::Input
 {
 public:
   Battery(const ControlState* level) : m_level(*level) {}
   std::string GetName() const override { return "Battery"; }
   ControlState GetState() const override { return m_level; }
+  bool IsDetectable() override { return false; }
```

Comment:
Seems weird that this method isn't `const`, but I guess thats from the base class. Any particular reason `Input::IsDetectable` isn't `const`?

Reply:
It should be. It's my my todo list. :P